[MRG] Warn when llvm + intel loaded at the same time

jeremiedbb opened the pull request:
Following the discussion in #40 and in particular [https://github.com/joblib/threadpoolctl/issues/40#issuecomment-545377697](https://github.com/joblib/threadpoolctl/issues/40#issuecomment-545377697)

both implementations of OpenMP are actually the same and are incompatible in the sense that they do not accept having multiple copies at the same time. See this error message from llvm
``OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program.``

(un)fortunately crashes should not occur very often outside of threadpoolctl since we are only able to create them by accessing the openmp api (get/set num threads, ...), not by mixing pranges and blas calls.

This PR adds the emission of a warning when trying to access threadpoolctl functionalities.

TODO:
 - [x] Add a page documenting all our discoveries about that and possible workarounds, and point the warning to that page

 - [x] test on macOS, especially the switch of threading layer for mkl

 - [x] add a test for the warning

 - [x] warn also in threadpool_limits

 - [x] What to do with all our clang ci jobs ?

ogrisel commented:
> What to do with all our clang ci jobs ?

Make sure we use numpy with openblas when testing under Linux with one of the clang compiled extension.

jeremiedbb commented:
> test on macOS, especially the switch of threading layer for mkl

there's no mkl init in numpy on macos. So mkl keeps intel threading layer even if llvm openmp is already loaded. This indicates that both libs are not incompatible as they are on linux. Moreover I haven't been able to reproduce any of the issue we encountered. It may be due (wild guess) to a difference in the dynamic loading process, but I really have no idea.

So only raise the warning on linux seems enough for now.

ogrisel commented:
This should be rebased on top of current master to use the new internal object oriented API.

jeremiedbb commented:
I rebased on top of master. Finally I did not add a test to check the warning since with this PR we'll avoid the mix libomp, libiomp on the ci. Codecov won't be happy once again :)

I added a page to document our OpenMP investigations. It's a first attempt, comments welcomed. Or even, feel free to edit directly.

ogrisel commented on multiple_openmp.md:
```diff
@@ -0,0 +1,35 @@
+with GCC, the most widely used C compiler).
+
+The only unrecoverable incompatibility we encountered is between libomp and
+libiomp, on Linux, manifested by crashes or deadlocks. It can happen even with
```
Maybe repeat the name of the matching compiler each time ... "between libomp (with clang/llvm) and libiomp (with ICC), on Linux, ...."

ogrisel commented on multiple_openmp.md:
```diff
@@ -0,0 +1,35 @@
+The only unrecoverable incompatibility we encountered is between libomp and
+libiomp, on Linux, manifested by crashes or deadlocks. It can happen even with
+the simplest OpenMP calls like getting the maximum number of threads that will
+be used in a subsequent parallel region. An explanation is that libomp is
```
We are not 100% sure this is the actual cause:

```suggestion
be used in a subsequent parallel region. A possible explanation is that libomp is
```

ogrisel commented on threadpoolctl.py:
```diff
@@ -523,6 +525,24 @@ def _check_prefix(self, library_basename, filename_prefixes):
+        prefixes = [module.prefix for module in self.modules]
+        msg = textwrap.dedent(
+            """
+            Found intel OpenMP ('libiomp') and llvm OpenMP ('libomp') loaded at
```
```suggestion
            Found Intel OpenMP ('libiomp') and LLVM OpenMP ('libomp') loaded at
```

ogrisel commented on threadpoolctl.py:
```diff
@@ -523,6 +525,24 @@ def _check_prefix(self, library_basename, filename_prefixes):
+        msg = textwrap.dedent(
+            """
+            Found intel OpenMP ('libiomp') and llvm OpenMP ('libomp') loaded at
+            the same time. Both libraries are known to be incompatible.
```
and this can cause random crashes or deadlocks on Linux when loaded in the same Python program.

ogrisel commented on multiple_openmp.md:
```diff
@@ -0,0 +1,35 @@
+threadpoolctl may crash your program in such a setting.
+
+Surprisingly, we never encountered this kind of issue on macOS, where this mix
+is the most frequent (Clang being the default C compiler on macOS).
```
We should have a section on possible workarounds:

- Building your OpenMP-enabled extensions with GCC (or ICC) instead of clang/LLVM under Linux.
- How to install numpy with openblas using conda from default channel or conda-forge.

ogrisel commented on .azure_pipeline.yml:
```diff
@@ -53,30 +53,38 @@ stages:
           CC_INNER_LOOP: 'clang-8'
-        # Linux environment with numpy from conda-forge channel
+        # Linux environment with numpy from conda-forge channel and
+        # heterogeneous OpenMP runtimes.
```
```suggestion
        # homogeneous OpenMP runtimes.
```

gcc inner & outer loops => homogeneous

jeremiedbb commented on tests/test_threadpoolctl.py:
```diff
@@ -361,3 +362,29 @@ def test_blis_threading_layer():
+    from ._openmp_test_helper import check_nested_openmp_loops  # noqa
+
+    # Trigger the import of numpy to potentially import Intel OpenMP via MKL
+    import numpy.linalg  # noqa
```
This is not necessary since we already import numpy for the whole module

ogrisel replied:
explicit is better than implicit :)

ogrisel commented:
Thank you very much @jeremiedbb. I think it's time for a release. WDYT?
